### PR TITLE
E4S Cray CI Stack

### PR DESCRIPTION
**Pull request by eugeneswalker**
Adding `E4S Cray` CI Stack

CC @scottwittenburg

### CODE REVIEW

**Comment by eugeneswalker** [comment]
All `E4S Cray` CI stack jobs are green!

The only blocker right now is `aws-pcluster-neoverse-n1` CI stack:
* [(specs) quantum-espresso/m4xxpxs 6.6 gcc@=12.3.0 linux-amzn2-neoverse_v1 AWS Packages](https://gitlab.spack.io/spack/spack/-/jobs/7081316)

**Comment by eugeneswalker** [comment]
@spackbot run pipeline

**Comment by eugeneswalker** [comment]
@spackbot run pipeline because of errors indicating that expected dependency binaries are missing from the cache for `aws-pcluster-skylake` stack. Hoping retry will render a better outcome...

**Comment by eugeneswalker** [comment]
Another `aws-pcluster-skylake` job is failing because of dependency binaries it expects to be in the cache but are missing, this time the job is for `intel-oneapi-mkl` and the missing binary is `intel-oneapi-tbb`:
* [(specs) intel-oneapi-mkl/k27p6if 2023.1.0 gcc@=12.3.0 linux-amzn2-icelake AWS Packages](https://gitlab.spack.io/spack/spack/-/jobs/7114287)
```
...
==> Error: No binary for intel-oneapi-tbb-2021.9.0-w7aicxeyfp5umxrqzh753uofmgzqfuus found when cache-only specified
==> Error: Failed to install intel-oneapi-tbb due to SystemExit: 1
==> Error: intel-oneapi-tbb-2021.9.0-w7aicxeyfp5umxrqzh753uofmgzqfuus: Package was not installed
==> Error: Installation request failed.  Refer to reported errors for failing package(s).
```
CC @scottwittenburg any ideas what's happening here?
@spackbot run pipeline

**Comment by scottwittenburg** [comment]
> CC @scottwittenburg any ideas what's happening here?

Ummm, I think I should not have allowed #37627 to be merged without some change to prevent what you're seeing here.  But that PR was rushed through due to the impending ISC deadline.  I'll see if I can come up with a way to work around this problem without having to revert that PR.

FYI @stephenmsachs I should have more insistent last week about [this bit](https://github.com/spack/spack/pull/37627#discussion_r1196811275).  For a lot of reasons, the policy we have adopted in the pipelines is that it is an error if a rebuild job can't find all it's dependencies in the binary cache.  Now the deliberate deletion of some built binaries from the aws-pcluster stack mirror is breaking pipelines and going to start affecting more PRs.

In some way this PR (adding the E4S Cray stack) faces a similar challenge.  Since we are not allowed to redistribute binaries built with cray tools, we had to create a separate, non-public bucket to store them.  This allows us to make sure some cray specs remain buildable, without allowing the binaries to be consumed publicly.  We could do something similar for the aws-pcluster stacks, but you have the additional constraint that you want the public to see *most* of the binaries built by your stack, so I think the idea of a separate, non-public bucket won't work for aws-pcluster.

**Comment by stephenmsachs** [comment]
Hi @scottwittenburg , I am terribly sorry to be causing such pain here. As far as I can see the errors in the pipelines here did not stem from the deliberate deletion in signing stage but an external deletion of the entire stage area. I had to ask for that as there were packages built with wrong relocation paths. I believe this should straighten itself out now.

The solution with the non-public bucket would of course be ideal. Maybe we can find something similar.

**Comment by scottwittenburg** [comment]
> I am terribly sorry to be causing such pain here.

Don't sweat it, it's actually my fault since you could not be expected to know the pipeline policy about never installing dependencies "from source" :sweat_smile: 

> the errors in the pipelines here did not stem from the deliberate deletion in signing stage but an external deletion of the entire stage area

I added a comment on the aws-pcluster PR that points out the line I believe makes the deliberate deletion of binaries an eventual error.

> The solution with the non-public bucket would of course be ideal. Maybe we can find something similar.

Creating a non-public bucket for the binaries for your stacks might be the easiest way forward, but would that be acceptable on your end?  If we do that, then using those binaries will require the user to have aws access credentials granting them read access.

**Comment by stephenmsachs** [comment]
> Creating a non-public bucket for the binaries for your stacks might be the easiest way forward, but would that be acceptable on your end? If we do that, then using those binaries will require the user to have aws access credentials granting them read access.

As discussed in Slack, a single non-public bucket would be contrary to what I am trying to achieve. I want to make the binaries which were linked against licensed libraries widely available and easy to consume for the inexperienced user.

**Comment by eugeneswalker** [comment]
@spackbot run pipeline

File: share/spack/gitlab/cloud_pipelines/stacks/e4s-cray/spack.yaml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+  ci:
+    pipeline-gen:
+    - build-job-remove:
+        image: no-image
+        tags: [spack, public]
+    - build-job:
+        tags:  [ "craype", "cce@15.0.1", "zen4" ]
+        script::
+        - - spack compiler find
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/stacks/e4s-cray/spack.yaml]: Is there are reason this isn't using the root configs for mapping and build script?

File: share/spack/gitlab/cloud_pipelines/stacks/e4s-cray/spack.yaml [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  - superlu-dist
+  - tau
+
+  mirrors: { "mirror": "s3://spack-binaries-cray/develop/e4s-cray" }
+
+  ci:
+    pipeline-gen:
+    - build-job-remove:
+        image: no-image
+        tags: [spack, public]
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/stacks/e4s-cray/spack.yaml]: Remove this?

**Review by kwryankrattiger — COMMENTED** [review]
Just a couple minor comments.

**Comment by eugeneswalker** [comment]
@spackbot rebuild everything

**Comment by eugeneswalker** [comment]
@spackbot run pipeline

**Comment by tldahlgren** [comment]
Looks like some CI stacks had pod not ready issues.

@spackbot run pipeline

**Comment by kwryankrattiger** [comment]
Looks like the cray CI is running now. hopefully we can get the rest to pass now.

**Comment by eugeneswalker** [comment]
@spackbot run pipeline

**Comment by eugeneswalker** [comment]
@spackbot run pipeline almost there, just a couple more pcluster jobs...

**Comment by eugeneswalker** [comment]
All CI is GREEN!

**Review by kwryankrattiger — APPROVED** [review]

**Pull request merged**